README.md
# Kopano Collection for Ansible

Documentation for the collection.

Managed Kopano's object, when server is  configured to use the [The DB Authentication Plugin](https://documentation.kopano.io/kopanocore_administrator_manual/configure_kc_components.html#the-db-authentication-plugin).  

## Code of Conduct

We follow the [Ansible Code of Conduct](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html) in all our interactions within this project.

If you encounter abusive behavior violating the [Ansible Code of Conduct](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html), please refer to the [policy violations](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html#policy-violations) section of the Code of Conduct for information on how to raise a complaint.

## Contributing

The content of this collection is made by [people](https://github.com/hlepesant/kopano/blob/main/CONTRIBUTORS) just like you, a community of individuals collaborating on making the world better through developing automation software.

We are actively accepting new contributors.

Any kind of contribution is very welcome.

You don't know how to start? Refer to our [contribution guide](https://github.com/hlepesant/kopano/blob/main/CONTRIBUTING.md)!

## Collection maintenance

The current maintainers (contributors with `write` or higher access) are listed in the [MAINTAINERS](https://github.com/hlepesant/kopano/blob/main/MAINTAINERS) file. If you have questions or need help, feel free to mention them in the proposals.

To learn how to maintain / become a maintainer of this collection, refer to the [Maintainer guidelines](https://github.com/hlepesant/kopano/blob/main/MAINTAINING.md).

It is necessary for maintainers of this collection to be subscribed to:

* The collection itself (the `Watch` button -> `All Activity` in the upper right corner of the repository's homepage).
* The "Changes Impacting Collection Contributors and Maintainers" [issue](https://github.com/hlepesant/kopano/issues).

They also should be subscribed to Ansible's [The Bullhorn newsletter](https://docs.ansible.com/ansible/devel/community/communication.html#the-bullhorn).

## Communication

Have a look to [CHANGELOG](https://github.com/hlepesant/kopano/blob/main/CHAGELOG.md).

For more information about communication, refer to the [Ansible Communication guide](https://docs.ansible.com/ansible/devel/community/communication.html).

## Governance

The process of decision making in this collection is based on discussing and finding consensus among participants.

Every voice is important and every idea is valuable. If you have something on your mind, create an issue or dedicated discussion and let's discuss it!

## Included content

- **Modules**:
  - [kopano_public_store](https://github.com/hlepesant/kopano/blob/main/docs/hlepesant.kopano.public_store_module.rst)
  - [kopano_company](https://github.com/hlepesant/kopano/blob/main/docs/hlepesant.kopano.company.rst)
  - [kopano_db_user](https://github.com/hlepesant/kopano/blob/main/docs/hlepesant.kopano.db.user.rst)
  - [kopano_db_group](https://github.com/hlepesant/kopano/blob/main/docs/hlepesant.kopano.db.group.rst)


## Tested with

### ansible-core

- 2.12

### Kopano Server

- kopano 8.7.0 on Debian 10 Buster

## External requirements

The Kopano modules rely on a [python3-kopano](https://packages.debian.org/buster/python3-kopano).

## Using this collection

### Installing the Collection from Ansible Galaxy

Before using the Kopano collection, you need to install it with the Ansible Galaxy CLI:

```bash
ansible-galaxy collection install hlepesant.kopano
```

You can also include it in a `requirements.yml` file and install it via `ansible-galaxy collection install -r requirements.yml`, using the format:

```yaml
---
collections:
  - name: hlepesant.kopano
```

Note that if you install the collection from Ansible Galaxy, it will not be upgraded automatically if you upgrade the Ansible package. To upgrade the collection to the latest available version, run the following command:

```bash
ansible-galaxy collection install hlepesant.kopano --upgrade
```

You can also install a specific version of the collection, for example, if you need to downgrade when something is broken in the latest version (please report an issue in this repository). Use the following syntax:

```bash
ansible-galaxy collection install hlepesant.kopano:==0.0.2
```

See [Ansible Using collections](https://docs.ansible.com/ansible/latest/user_guide/collections_using.html) for more details.

## Licensing

<!-- Include the appropriate license information here and a pointer to the full licensing details. If the collection contains modules migrated from the ansible/ansible repo, you must use the same license that existed in the ansible/ansible repo. See the GNU license example below. -->

GNU General Public License v3.0 or later.

See [LICENSE](https://www.gnu.org/licenses/gpl-3.0.txt) to see the full text.




CHANGELOG.md
# Kopano Collection for Ansible Changelog

## v0.0.5
 - Fix typo

## v0.0.4
 - Fix doc extension

## v0.0.3
 - Fix doc path


## v0.0.2
 - Fix namespace


## v0.0.1
 - First release




plugins/modules/kopano_company.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright: (c) 2022, Hugues Lepesant <hugues@lepesant.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)


from __future__ import absolute_import, division, print_function
__metaclass__ = type

ANSIBLE_METADATA = {
    'metadata_version': '0.1',
    'status': ['preview'],
    'supported_by': 'hlepesant'
}

DOCUMENTATION = r'''
---
module: kopano_company

short_description: Create a kopano company

description:
  - Create a kopano company

References : 
  - https://documentation.kopano.io/kopanocore_administrator_manual/special_kc_configurations.html#multi-tenancy-configurations

author: Hugues Lepesant (@hlepesant)
version: 0.1
options:

'''

EXAMPLES = r'''
- name: Create Kopano Company
  hlepesant.kopano.kopano_company:
    name: Zarafa
    state: present
'''

RETURN = '''#'''

from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils._text import to_native

from ansible_collections.hlepesant.kopano.plugins.module_utils.kopano_common import (
    missing_required_lib,
    kopano_found,
    E_IMP_ERR,
    kopano_common_argument_spec,
    KopanoHelpers,
    NotFoundError
)


def run_module():

  argument_spec = kopano_common_argument_spec()
  argument_spec.update(
    name=dict(type='str', required=True),
    state=dict(type='str', default='present'),
  )

  module = AnsibleModule(
    argument_spec=argument_spec,
    supports_check_mode=True,
  )

  if not kopano_found:
    module.fail_json(msg=missing_required_lib('kopano'),
      exception=E_IMP_ERR)

  name = module.params['name']
  state = module.params['state']

  try:
    kopano = KopanoHelpers(module)
    k = kopano.connect()

    if not k.multitenant:
      module.exit_json(changed=False, msg="Unable to create company {0}: action not supported by server in single-tenancy support. ".format(name))

    _company = k.get_company(name)

    if _company is None:
      if state == 'present':
        k.create_company(name)
        module.exit_json(changed=True, msg="The company {0} was created.".format(name))
      else:
        module.exit_json(changed=False, msg="The company {0} does not exists.".format(name))
    else:
      if state == 'absent':
        k.delete(_company)
        # k.remove_company(name)
        module.exit_json(changed=True, msg="The company {0} was deleted.".format(name))
      else:
        module.exit_json(changed=False, msg="The company {0} already exists.".format(name))

  except Exception as excep:
    module.fail_json(msg='Kopano error: %s' % to_native(excep))


def main():
    run_module()

if __name__ == '__main__':
    main()


plugins/modules/kopano_db_group.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright: (c) 2022, Hugues Lepesant <hugues@lepesant.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)


from __future__ import absolute_import, division, print_function
from ansible_collections.hlepesant.kopano.plugins.module_utils.kopano_common import (
    missing_required_lib,
    kopano_found,
    E_IMP_ERR,
    kopano_common_argument_spec,
    KopanoHelpers,
    NotFoundError
)
from ansible.module_utils._text import to_native
from ansible.module_utils.basic import AnsibleModule
__metaclass__ = type

ANSIBLE_METADATA = {
    'metadata_version': '0.1',
    'status': ['preview'],
    'supported_by': 'hlepesant'
}

DOCUMENTATION = r'''
---
module: kopano_db_group

short_description: Create a kopano group

description:
  - Create a kopano group

References :
  - https://documentation.kopano.io/kopanocore_administrator_manual/user_management.html#groups

author: Hugues Lepesant (@hlepesant)
version: 0.1
options:

'''

EXAMPLES = r'''
- name: Create Kopano Group
  hlepesant.kopano.kopano_db_group:
    name: support
    email: support@zarafa.com
    members:
      - john
    state: present
'''

RETURN = '''#'''


def get_group(name):
    try:
        group = k.group(name, False)
    except kopano.NotFoundError:
        return None
    return group

def db_group_members(group, target):
    _change = False
    _members = group.users()
    mb = []
    for member in _members:
       mb.append(member.name)

    mb.sort()
    target.sort()

    to_remove = list(set(mb) - set(target))
    to_add = list(set(target) - set(mb))

    for username in to_remove:
        user = k.get_user(username)
        if user is not None:
            group.remove_user(user)
            _change = True

    for username in to_add:
        user = k.get_user(username)
        if user is not None:
            group.add_user(user)
            _change = True

    return _change

def db_group_send_as(group, target):
    _change = False
    _sendas = group.send_as()
    sa = []
    for sendas in _sendas:
       sa.append(sendas.name)

    sa.sort()
    target.sort()

    to_remove = list(set(sa) - set(target))
    to_add = list(set(target) - set(sa))

    for username in to_remove:
        user = k.get_user(username)
        if user is not None:
            group.remove_send_as(user)
            _change = True

    for username in to_add:
        user = k.get_user(username)
        if user is not None:
            group.add_send_as(user)
            _change = True

    return _change

def run_module():
    global k

    argument_spec = kopano_common_argument_spec()
    argument_spec.update(
        name=dict(type='str', required=True),
        fullname=dict(type='str', required=False, default=''),
        email=dict(type='str', required=False, default=''),
        hidden=dict(type='bool', default=False),
        members=dict(type='list', required=False, default=[]),
        send_as=dict(type='list', required=False, default=[]),
        state=dict(type='str', required=False, default='present'),
    )

    module = AnsibleModule(
        argument_spec=argument_spec,
        supports_check_mode=True,
    )

    result = dict(
        changed=False,
        original_message='',
        message=''
    )
    result['original_message'] = module.params['name']
    result['message'] = ""

    if not kopano_found:
        module.fail_json(msg=missing_required_lib('kopano'),
                         exception=E_IMP_ERR)

    name = module.params['name']
    fullname = module.params['fullname']
    email = module.params['email']
    state = module.params['state']
    hidden = module.params['hidden']
    members = module.params['members']
    send_as = module.params['send_as']

    try:
        kopano = KopanoHelpers(module)
        k = kopano.connect()

        g = get_group(name)

        if g is None:
            if state == "present":
                k.create_group(name, fullname, email, hidden, None)
                result['message'] += "The group {0} was created. ".format(name)
                result['changed'] = True

        if state == "absent":
            k.remove_group(g.name)
            result['message'] = "The group {0} was deleted.".format(name)
            result['changed'] = True
            module.exit_json(**result)

        # we need to refresh the group object
        # g = get_group(name)
        g = k.group(name, False)

        if ( db_group_members(g, members) ):
            result['message'] += "Members of group {0} has changed. ".format(name)
            result['changed'] = True

        if( db_group_send_as(g, send_as) ):
            result['message'] += "SendAs of group {0} has changed.".format(name)
            result['changed'] = True

        module.exit_json(**result)

    except Exception as excep:
        module.fail_json(msg='Kopano error: %s' % to_native(excep))


def main():
    run_module()


if __name__ == '__main__':
    main()


plugins/module_utils/kopano_common.py
from __future__ import absolute_import, division, print_function
__metaclass__ = type
from ansible.module_utils.basic import AnsibleModule, missing_required_lib
from ansible.module_utils.six.moves import configparser

import traceback
import sys

kopano_found = False
E_IMP_ERR = None
NotFoundError = None
helpers = None

try:
  import kopano
  kopano_found = True
except ImportError:
  E_IMP_ERR = traceback.format_exc()
  kopano_found = False


def kopano_common_argument_spec():
  """
  Returns a dict containing common options shared across the kopano modules
  """
  options = dict(
    server_socket=dict(type='str', default=None),
    sslkey_file=dict(type='str', default=None),
    sslkey_pass=dict(type='str', default=None, no_log=True),
    # config=dict(type='str', default=None),
    # auth_user=dict(type='str', default=None),
    # auth_pass=dict(type='str', default=None, no_log=True),
    # options=dict(type='str', default=None),
    # parse_args=dict(type='str', default=False),
  )
  return options

class KopanoHelpers():
  """
  Class containing helper functions for Kopano modules
  """

  def __init__(self, module):
    self.module = module

  def use_default_location(self, module):
    if module.params['server_socket'] is not None:
      return False
    return True

  def connect(self):
    default_location = self.use_default_location(self.module)

    if default_location:
      k = kopano.Server()
    else:
      k = kopano.server(
        module.params['server_socket'],
        module.params['sslkey_file'],
        module.params['sslkey_pass']
      )

    return k
    


plugins/modules/kopano_public_store.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright: (c) 2022, Hugues Lepesant <hugues@lepesant.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)


from __future__ import absolute_import, division, print_function
__metaclass__ = type

ANSIBLE_METADATA = {
    'metadata_version': '0.1',
    'status': ['preview'],
    'supported_by': 'hlepesant'
}

DOCUMENTATION = r'''
---
module: kopano_public_store

short_description: Create a kopano public store

description:
  - Create a kopano public store

references: 
  - https://documentation.kopano.io/kopanocore_administrator_manual/user_management.html#public-folder

author: Hugues Lepesant (@hlepesant)
version_
options:

'''

EXAMPLES = r'''
- name: Create Kopano public store
  hlepesant.kopano.kopano_public_store:
    state: present
'''

RETURN = '''#'''

from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils._text import to_native

from ansible_collections.hlepesant.kopano.plugins.module_utils.kopano_common import (
    missing_required_lib,
    kopano_found,
    E_IMP_ERR,
    kopano_common_argument_spec,
    KopanoHelpers,
    NotFoundError
)

def run_module():

    argument_spec = kopano_common_argument_spec()
    argument_spec.update(
        state=dict(type='str', required=False, default='present')
    )

    module = AnsibleModule(
        argument_spec=argument_spec,
        supports_check_mode=True,
    )

    result = dict(
        changed=False,
        original_message='',
        message=''
    )
    result['message'] = ''

    if not kopano_found:
        module.fail_json(msg=missing_required_lib('kopano'), exception=E_IMP_ERR)


    try:
        kopano = KopanoHelpers(module)
        k = kopano.connect()

        if k.multitenant:
            module.fail_json(msg="Kopano is not running a multi tenant environment.")

        _public_store = k.public_store
        state = module.params['state']
        
        if (_public_store is None):
            if state == "present":
                k.create_public_store()
                result['message'] += "The public store was created."
                result['changed'] = True
        else:
            result['message'] += "The public store already exists."
            result['changed'] = False
        
        if state == "absent":
            result['message'] += "The public store can not be deleted."
            result['changed'] = False
    
        module.exit_json(**result)

    except Exception as excep:
        module.fail_json(msg='Kopano error: %s' % to_native(excep))


def main():
    run_module()

if __name__ == '__main__':
    main()


plugins/modules/kopano_db_user.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright: (c) 2022, Hugues Lepesant <hugues@lepesant.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)


from __future__ import absolute_import, division, print_function
__metaclass__ = type

ANSIBLE_METADATA = {
    'metadata_version': '0.2',
    'status': ['preview'],
    'supported_by': 'hlepesant'
}

DOCUMENTATION = r'''
---
module: kopano_company

short_description: Create a kopano user

description:
  - Create a kopano user

References : 
  - https://documentation.kopano.io/kopanocore_administrator_manual/user_management.html#users-management-with-db-plugin

author: Hugues Lepesant (@hlepesant)
version: 0.1
options:

'''

EXAMPLES = r'''
- name: Create Kopano User
  hlepesant.kopano.kopano_db_user:
    name: john
    password: ahTon1ohYo8u
    email: john.doe@zarafa.com
    fullname: John Doe
    administrator: false
    state: present

- name: Create Kopano User and overwrite quota
  hlepesant.kopano.kopano_db_user:
    name: axel
    password: ahTon1erTo8u
    email: axel.doe@zarafa.com
    fullname: Axel Doe
    administrator: false
    quota_use_default: false
    quota_hard: 200 mb
    state: present

- name: Create Kopano User and overwrite and set all quota
  hlepesant.kopano.kopano_db_user:
    name: axel
    password: ahTon1erTo8u
    email: axel.doe@zarafa.com
    fullname: Axel Doe
    administrator: false
    quota_use_default: false
    quota_hard: 200 mb
    quota_soft: 195 mb
    quota_warn: 190 mb
    state: present
'''

RETURN = '''#'''

from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils._text import to_native

from ansible_collections.hlepesant.kopano.plugins.module_utils.kopano_common import (
    missing_required_lib,
    kopano_found,
    E_IMP_ERR,
    kopano_common_argument_spec,
    KopanoHelpers,
    NotFoundError
)

from kopano.utils import (
    bytes_to_human,
    human_to_bytes
)

# Send AS
def db_user_send_as(user, target):
    _change = False
    _sendas = user.send_as()
    sa = []
    for sendas in _sendas:
       sa.append(sendas.name)

    sa.sort()
    target.sort()

    to_remove = list(set(sa) - set(target))
    to_add = list(set(target) - set(sa))

    for username in to_remove:
        _user = k.get_user(username)
        if _user is not None:
            user.remove_send_as(_user)
            _change = True

    for username in to_add:
        _user = k.get_user(username)
        if _user is not None:
            user.add_send_as(_user)
            _change = True

    return _change

# Quotas
def db_user_quota(user, quota_use_default, quota_hard, quota_soft = None, quota_warn = None):

    _change = False
    soft_factor = 0.95
    warn_factor = 0.90

    if user.quota.use_default and quota_use_default:
        _change = False

    if not user.quota.use_default and quota_use_default:
        user.quota.update(
            use_default = True
        )
        _change = True

    if not quota_use_default:
        # hard quota (must be defined)
        quota_hard_b = human_to_bytes(quota_hard)

        # soft quota
        if quota_soft is None:
            quota_soft = bytes_to_human(quota_hard_b * soft_factor)
     
        quota_soft_b = human_to_bytes(quota_soft)

        # warning quota
        if quota_warn is None:
            quota_warn = bytes_to_human(quota_hard_b * warn_factor)
    
        quota_warn_b = human_to_bytes(quota_warn)
        
        if quota_hard_b != user.quota.hard_limit or quota_soft_b != user.quota.soft_limit or quota_warn_b != user.quota.warning_limit:
        
            user.quota.update(
                use_default   = False,
                warning_limit = quota_warn_b,
                soft_limit    = quota_soft_b,
                hard_limit    = quota_hard_b,
            )
            _change = True

    return _change


def run_module():
    global k

    argument_spec = kopano_common_argument_spec()
    argument_spec.update(
        name=dict(type='str', required=True),
        email=dict(type='str', required=True),
        password=dict(type='str', required=True, no_log=True),
        company=dict(type='str', required=False, default=None),
        fullname=dict(type='str', required=True),
        administrator=dict(type='bool', required=False, default=False),
        update_password=dict(type='bool', required=False, default=False, no_log=True),
        send_as=dict(type='list', required=False, default=[]),
        quota_use_default=dict(type='bool', required=False, default=True),
        quota_hard=dict(type='str', required=False, default=None),
        quota_soft=dict(type='str', required=False, default=None),
        quota_warn=dict(type='str', required=False, default=None),
        state=dict(type='str', required=False, default='present'),
    )

    module = AnsibleModule(
        argument_spec=argument_spec,
        supports_check_mode=True,
        required_if=[('quota_use_default', False, ['quota_hard'])],
        # required_together=[('quota_soft', 'quota_warn'])],
    )

    result = dict(
        changed=False,
        original_message='',
        message=''
    )
    result['original_message'] = module.params['name']
    result['message'] = ""

    if not kopano_found:
        module.fail_json(msg=missing_required_lib('kopano'), exception=E_IMP_ERR)
    
    name = module.params['name']
    email = module.params['email']
    password = module.params['password']
    company = module.params['company']
    fullname = module.params['fullname']
    administrator = module.params['administrator']
    update_password = module.params['update_password']
    send_as = module.params['send_as']
    quota_use_default = module.params['quota_use_default']
    quota_hard = module.params['quota_hard']
    quota_soft = module.params['quota_soft']
    quota_warn = module.params['quota_warn']
    state = module.params['state']

    _admin_level = 0
    _company = None

    try:
        kopano = KopanoHelpers(module)
        k = kopano.connect()

        if k.multitenant and company is not None:
            _company = k.get_company(company)
        
        if k.multitenant and administrator:
            _admin_level = 1
        
        _user = k.get_user(name)
        
        if _user is None:
            if state == 'present':
                k.create_user(name, email, password, _company, fullname, True)
                result['changed'] = True
                result['message'] = "The user {0} was created.".format(name)
            else:
                result['changed'] = False
                result['message'] = "The user {0} does not exists.".format(name)
        
            module.exit_json(**result)

        # user found
        if state == 'absent':
            k.remove_user(_user)
            result['changed'] = True
            result['message'] = "The user {0} was deleted.".format(name)
        
        else:
        
            if _user.email != email:
                _user.email = email
                result['changed'] = True
                result['message'] += "The email for user {0} was udpated.".format(name)
            
            if _user.fullname != fullname:
                _user.fullname = fullname
                result['changed'] = True
                result['message'] += "The fullname for user {0} was udpated.".format(name)
            
            if update_password:
                _user.password = password
                result['changed'] = True
                result['message'] += "The password for user {0} was udpated.".format(name)

            if( db_user_send_as(_user, send_as) ):
                result['message'] += "SendAs of user {0} has changed.".format(name)
                result['changed'] = True

            # quotas
            quota_changed = False
            quota_changed = db_user_quota(_user, quota_use_default, quota_hard, quota_soft, quota_warn)

            if quota_changed:
                result['message'] += "Quota configuration for user {0} has changed.".format(name)
                result['changed'] = True

            
        module.exit_json(**result)
    
    except Exception as excep:
        module.fail_json(msg='Kopano error: %s' % to_native(excep))


def main():
    run_module()

if __name__ == '__main__':
    main()
